lib/gears/reactive.lua

```diff
@@ -0,0 +1,647 @@
+end
+
+local function proxy_lesser(a, b)
+    a, b = get_upvalue(a), get_upvalue(b)
+
+    return a < b
+end
+
+local function proxy_lesser_equal(a, b)
+    a, b = get_upvalue(a), get_upvalue(b)
+
+    return a <= b
+end
+
+local function proxy_tostring(o)
+    return tostring(get_upvalue(o))
+end
+
+-- Wrap tables and functions into a proxy object.
+create_proxy = function(input, origin, getter)
+    local t = type(input)
+    local is_call = is_callable(input)
+
+    -- Everything but the tables are immutable.
+    if t ~= "table" and not is_call then return input, false end
+
+    -- Remove any foreign proxy.
+    if t ~= "function" and input._reactive and input._reactive.origin ~= origin then
+        input = get_upvalue(input)
+    end
+
+    return setmetatable({_reactive = gen_proxy_md(input, origin, getter)}, {
+        __index    = proxy_index,
+        __newindex = proxy_newindex,
+        __call     = proxy_call,
+        __eq       = proxy_equal,
+        __lt       = proxy_lesser,
+        __le       = proxy_lesser_equal,
+        __tostring = proxy_tostring
+    }), true
+end
+
+-- Create a "fake" upvalue ref because otherwise the sandbox could
+-- "leak" and set upvalue in actual code outside of the function.
+-- By "leak", I mean the proxy would become visible from outside
+-- of the sandbox by the reactive expression "siblings" sharing the
+-- same environment.
+local function copy_upvalue_reference(loaded, idx, value)
+    -- Make sure it has a fresh upvalue reference. Something that
+    -- we are sure cannot be shared with something else.
+    local function fake_upvalue_env()
+        return value
+    end
+
+    debug.upvaluejoin(loaded, idx, fake_upvalue_env, 1) --luacheck: ignore
+end
+
+-- Create a meta-getter for the original `fct` upvalue at index `idx`.
+-- This will allow all the other code to dynamically get a "current" version
+-- of the upvalue rather than something from the cache. It will also avoid
+-- having to store a strong reference to the original value.
+local function create_upvalue_getter(fct, idx)
+    local placeholder = nil
+
+    local function fake_upvalue_getter()
+        return placeholder
+    end
+
+    -- Using this, each time `fake_upvalue_getter` is called, it
+    -- will return the current upvalue. This means we don't have to
+    -- cache it. Thus, the cache cannot get outdated. The drawback if
+    -- that if the type changes from an object to a primitive, it
+    -- will explode.
+    debug.upvaluejoin(fake_upvalue_getter, 1, fct, idx) --luacheck: ignore
+
+    return fake_upvalue_getter
+end
+
+-- Create a copy of the function and replace it's environment.
+local function sandbox(fct, env, vars, values)
+    -- Serialize to make sure it cannot affect the real
+    -- upvalues anymore. Without this setting the proxy actually
+    -- affects other function with the same uplavue environment.
+    local dump   = string.dump(fct)
+    local loaded = load(dump, nil, nil, env)
```

Comment:
Ahhh... that's why you are not messing with functions outside of... things. I bet this cryptic sentence is not understandable, but something makes now a bit more sense to me.

Reply:
> I bet this cryptic sentence is not understandable,

Rewritten